src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphProjectCacheUtility.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+
+namespace NuGet.PackageManagement
+{
+    public static class DependencyGraphProjectCacheUtility
+    {
+        /// <summary>
+        /// Creates an index of the project unique name to the cache entry.
+        /// The cache entry contains the project and the closure of project.json files.
+        /// </summary>
+        public static async Task<Dictionary<string, DependencyGraphProjectCacheEntry>>
+            CreateDependencyGraphProjectCache(
+                IEnumerable<IDependencyGraphProject> projects,
+                ExternalProjectReferenceContext context)
+        {
+            var cache = new Dictionary<string, DependencyGraphProjectCacheEntry>();
+
+            // Find all project closures
+            foreach (var project in projects)
+            {
+                // Get all project.json file paths in the closure
+                var closure = await project.GetProjectReferenceClosureAsync(context);
+
+                var files = new HashSet<string>(StringComparer.Ordinal);
```

Comment:
This was pre-existing (I just moved the code around). That being said, it makes sense to me given that this is in NuGet.Core and could therefore (potentially) be consumed from a case-sensitive file system. Also, this case sensitivity means that the `CacheHasChanges` method will be more sensitive to changes, which seems like a safer flow than ignoring case changes in files.